### PR TITLE
Add type annotations to schema.utils

### PR DESCRIPTION
**Pull request by RobertoPrevato**

### CODE REVIEW

File: edb/schema/utils.py [hunk 1]
```diff
@@ -391,17 +452,25 @@ def find_item_suggestions(
...
 def enrich_schema_lookup_error(
-        error, item_name, modaliases, schema, *,
-        item_types, suggestion_limit=3, name_template=None,
-        collection=None, condition=None, context=None):
+        error: errors.EdgeDBError,
+        item_name: Union[sn.SchemaName, str, None],
+        modaliases: Mapping[Optional[str], str],
+        schema: s_schema.Schema,
+        *,
+        item_types: Tuple[so.Type_T, ...] = (),
```
Comment (elprans) [inline edb/schema/utils.py]: Likewise, `so.ObjectMeta` is a better annotation here.

File: edb/schema/utils.py [hunk 2]
```diff
@@ -320,21 +371,30 @@ def get_nq_name(schema, item) -> str:
 
 
 def find_item_suggestions(
-        name, modaliases, schema, *, item_types=(), limit=3,
-        collection=None, condition=None):
+        name: Union[sn.SchemaName, str, None],
```
Comment (elprans) [inline edb/schema/utils.py]: `sn.SchemaName` is a subclass of `str`, so I wonder if `Optional[str]` is enough here.

File: edb/schema/utils.py [hunk 3]
```diff
@@ -391,17 +452,25 @@ def find_item_suggestions(
 
 
 def enrich_schema_lookup_error(
-        error, item_name, modaliases, schema, *,
-        item_types, suggestion_limit=3, name_template=None,
-        collection=None, condition=None, context=None):
+        error: errors.EdgeDBError,
+        item_name: Union[sn.SchemaName, str, None],
```
Comment (elprans) [inline edb/schema/utils.py]: `sn.SchemaName` is a subclass of `str`, so I wonder if `Optional[str]` is enough here.

File: edb/schema/utils.py [hunk 4]
```diff
@@ -438,60 +507,71 @@ def ensure_union_type(
...
     *,
     opaque: bool = False,
     module: Optional[str] = None,
 ) -> Tuple[s_schema.Schema, s_types.Type]:
+    from . import schema as s_schema
 
     schema, union, _ = ensure_union_type(
         schema, types, opaque=opaque, module=module)
 
+    assert isinstance(schema, s_schema.Schema)
```
Comment (elprans) [inline edb/schema/utils.py]: This `assert` seems redundant

File: edb/schema/utils.py [hunk 5]
```diff
@@ -71,23 +70,33 @@ def ast_to_typeref(
...
+        assert isinstance(node.maintype, qlast.ObjectRef)
         coll = s_types.Collection.get_class(node.maintype.name)
 
-        if issubclass(coll, s_abc.Tuple):
-            subtypes = collections.OrderedDict()
+        if issubclass(coll, s_types.Tuple):
+            # Note: if we used abc Tuple here, then we would need anyway
+            # to assert it is an instance of s_types.Tuple to make mypy happy
+            # (rightly so, because later we use from_subtypes method)
+            # assert isinstance(coll, s_types.Tuple)
```
Comment (1st1) [inline edb/schema/utils.py]: Is this some commented out code?

File: edb/schema/utils.py [hunk 6]
```diff
@@ -438,60 +507,71 @@ def ensure_union_type(
...
-            raise _union_error(schema, components)
+            if common_type is None:
+                raise _union_error(schema, components_list)
+            else:
+                uniontype = common_type
     else:
         schema, uniontype, created = s_objtypes.get_or_create_union_type(
-            schema, components=components, opaque=opaque, module=module)
+            schema,
+            components=cast(Iterable[s_objtypes.ObjectType],
```
Comment (1st1) [inline edb/schema/utils.py]: It's a bit weird that you had to put a cast here. What type is `components_list` without a cast?

File: edb/schema/utils.py (hunk 5 above)
Comment (RobertoPrevato) [reply edb/schema/utils.py]: Right - I am removing the last line.

File: edb/schema/utils.py (hunk 4 above)
Comment (RobertoPrevato) [reply edb/schema/utils.py]: Right - I am removing the last line.

File: edb/schema/utils.py (hunk 6 above)
Comment (RobertoPrevato) [reply edb/schema/utils.py]: Right :+1:

File: edb/schema/utils.py (hunk 3 above)
Comment (RobertoPrevato) [reply edb/schema/utils.py]: Right. :+1:

File: edb/schema/utils.py (hunk 2 above)
Comment (RobertoPrevato) [reply edb/schema/utils.py]: Right. :+1:

**Comment by 1st1** [comment]
Congrats on first merged PR! 🎉

**Pull request merged**